curp/src/client.rs

```diff
@@ -1 +1,139 @@
+use std::{fmt::Debug, marker::PhantomData, net::SocketAddr, sync::Arc, time::Duration};
 
+use futures::{stream::FuturesUnordered, StreamExt};
+use madsim::net::Endpoint;
+use tracing::{error, log::warn};
+
+use crate::{
+    cmd::Command,
+    message::{Propose, ProposeError},
+};
+
+/// Propose request default timeout
+static PROPOSE_TIMEOUT: Duration = Duration::from_secs(1);
+
+/// Protocol client
+pub struct Client<C: Command> {
+    /// Client endpoint to send requests
+    ep: Arc<Endpoint>,
+    /// Leader address
+    #[allow(dead_code)] // should use in propose synced
+    leader: SocketAddr,
+    /// All servers addresses including leader address
+    addrs: Vec<SocketAddr>,
+    /// To keep Command type
+    phatom: PhantomData<C>,
+}
+
+impl<C> Debug for Client<C>
+where
+    C: Command,
+{
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Client")
+            .field("addrs", &self.addrs)
+            .field("phatom", &self.phatom)
+            .finish()
+    }
+}
+
+impl<C> Client<C>
+where
+    C: Command + 'static,
+{
+    /// Create a new protocol client based on the addresses
+    #[inline]
+    pub async fn new(leader: &SocketAddr, addrs: &[SocketAddr]) -> Option<Self> {
+        let ep = Endpoint::bind("127.0.0.1:0")
+            .await
+            .map_err(|e| {
+                error!("initial endpoint error: {e}");
+            })
+            .ok()?;
+
+        Some(Self {
+            ep: Arc::new(ep),
+            leader: *leader,
+            addrs: addrs.to_owned(),
+            phatom: PhantomData::<C>::default(),
+        })
+    }
+
+    /// Propose the request to servers
+    /// # Errors
+    ///   `ProposeError::ExecutionError` if execution error is met
+    #[inline]
+    pub async fn propose(&self, cmd: C) -> Result<C::ER, ProposeError> {
+        let ft = self.addrs.len() / 2;
+        let rpcs = self.addrs.iter().map(|addr| {
+            let cmd = cmd.clone();
+            self.ep
+                .call_timeout(*addr, Propose::new(cmd), PROPOSE_TIMEOUT)
+        });
+
+        let mut rpcs: FuturesUnordered<_> = rpcs.collect();
+
+        let mut ok_cnt: usize = 0;
+        let mut max_term = 0;
+        let mut execute_result: Option<C::ER> = None;
+
+        while ok_cnt
+            >= (ft
+                .wrapping_add(ft.wrapping_add(1).wrapping_div(2))
+                .wrapping_add(1))
+            && execute_result.is_some()
```

Comment:
Is this logic reversed?

Reply:
What is the reason of using wrapping_add instead of overflowing_add?

Reply:
Yes, good catch. we can make sure ft the result is less than usize limit for that `ft = peer_cnt / 2`.